### PR TITLE
1.15 Update

### PR DESCRIPTION
**Pull request by arthurbambou**
Fixes crash on snapshot more recent than 19w38

### CODE REVIEW

File: src/main/java/com/swordglowsblue/artifice/mixin/MixinMinecraftClient.java [hunk 1]
```diff
@@ -5,41 +5,46 @@
...
 
-    @SuppressWarnings("unchecked")
-    @Inject(method = "<init>", at = @At("RETURN"))
-    private void registerPackCreator(CallbackInfo cbi) {
-        this.resourcePackContainerManager.addCreator(new ResourcePackCreator() {
-            public <T extends ResourcePackContainer> void registerContainer(Map<String, T> packs, ResourcePackContainer.Factory<T> factory) {
+    @Redirect(method = "<init>", at = @At(value = "NEW", target = "net/minecraft/client/tutorial/TutorialManager"))
+    private TutorialManager registerPackCreator(MinecraftClient minecraftClient_1) {
+        this.resourcePackManager.registerProvider(new ResourcePackProvider() {
+            public <T extends ResourcePackProfile> void register(Map<String, T> packs, ResourcePackProfile.Factory<T> factory) {
```
Comment (swordglowsblue) [inline src/main/java/com/swordglowsblue/artifice/mixin/MixinMinecraftClient.java]: I'm curious - why the change to a redirect here? Is there a reasonable way to make this an injection instead for compatibility reasons?

File: src/main/java/com/swordglowsblue/artifice/mixin/MixinMinecraftClient.java [hunk 2]
```diff
@@ -5,41 +5,46 @@
...
-    private void enableNonOptional(CallbackInfo cbi) {
-        this.resourcePackContainerManager.getDisabledContainers().forEach(c -> {
-            if(c instanceof ArtificeResourcePackContainer && !((ArtificeResourcePackContainer)c).isOptional())
-                this.resourcePackContainerManager.getEnabledContainers().add(c);
+    @Redirect(method = "<init>", at = @At(value = "INVOKE", target = "net/minecraft/resource/ResourcePackManager.scanPacks()V"))
+    private void enableNonOptional(ResourcePackManager resourcePackContainerManager) {
+        resourcePackContainerManager.scanPacks();
+        resourcePackContainerManager.getDisabledProfiles().forEach(c -> {
+
+            if(c instanceof ArtificeResourcePackContainer && !((ArtificeResourcePackContainer)c).isOptional()) resourcePackContainerManager.getEnabledProfiles().add((ClientResourcePackProfile) c);
```
Comment (swordglowsblue) [inline src/main/java/com/swordglowsblue/artifice/mixin/MixinMinecraftClient.java]: Needs formatting fixes:
```
if(c instanceof ArtificeResourcePackContainer && !((ArtificeResourcePackContainer)c).isOptional())
    resourcePackContainerManager.getEnabledProfiles().add((ClientResourcePackProfile) c);
```

File: src/main/java/com/swordglowsblue/artifice/mixin/MixinMinecraftClient.java [hunk 3]
```diff
@@ -5,41 +5,46 @@
...
+        return new TutorialManager(minecraftClient_1);
     }
 
-    @Inject(method = "init", at = @At(value = "INVOKE",
-            target = "net/minecraft/resource/ResourcePackContainerManager.callCreators()V"))
-    private void enableNonOptional(CallbackInfo cbi) {
-        this.resourcePackContainerManager.getDisabledContainers().forEach(c -> {
-            if(c instanceof ArtificeResourcePackContainer && !((ArtificeResourcePackContainer)c).isOptional())
-                this.resourcePackContainerManager.getEnabledContainers().add(c);
+    @Redirect(method = "<init>", at = @At(value = "INVOKE", target = "net/minecraft/resource/ResourcePackManager.scanPacks()V"))
```
Comment (swordglowsblue) [inline src/main/java/com/swordglowsblue/artifice/mixin/MixinMinecraftClient.java]: Again, curious why the redirect. Would like to avoid them if reasonably possible.

File: src/testmod/java/com/swordglowsblue/artifice/test/ArtificeTestMod.java [hunk 4]
```diff
@@ -32,8 +32,10 @@ public void onInitialize() {
 
     public void onInitializeClient() {
         ArtificeResourcePack resourcePack = Artifice.registerAssets("artifice:testmod", pack -> {
+            pack.setVisible();
             pack.setDisplayName("Artifice Test Resources");
             pack.setDescription("Resources for the Artifice test mod");
+            pack.setVisible();
```
Comment (swordglowsblue) [inline src/testmod/java/com/swordglowsblue/artifice/test/ArtificeTestMod.java]: Why does this need to be changed, and why the duplicate afterwards? Keep in mind that `setOptional` also calls `setVisible`, if you're considering this for coverage purposes.

File: src/main/java/com/swordglowsblue/artifice/mixin/MixinMinecraftClient.java (hunk 3 above)
Comment (arthurbambou) [reply src/main/java/com/swordglowsblue/artifice/mixin/MixinMinecraftClient.java]: You can't inject at a specific point on constructor, only `@At(value = "RETURN")` isn't allowed.

File: src/main/java/com/swordglowsblue/artifice/mixin/MixinMinecraftClient.java (hunk 1 above)
Comment (arthurbambou) [reply src/main/java/com/swordglowsblue/artifice/mixin/MixinMinecraftClient.java]: same here and init method has been merged in constructor, it didn't worked when at RETURN.

File: src/testmod/java/com/swordglowsblue/artifice/test/ArtificeTestMod.java (hunk 4 above)
Comment (arthurbambou) [reply src/testmod/java/com/swordglowsblue/artifice/test/ArtificeTestMod.java]: Removed

**Comment by arthurbambou** [comment]
did you see the change I made? Did I answer your questions?

**Comment by swordglowsblue** [comment]
Sorry, I'm a bit swamped with IRL stuff right now. I'll get to this when I can.

**Comment by arthurbambou** [comment]
No problem :-)

File: src/main/java/com/swordglowsblue/artifice/mixin/MixinMinecraftClient.java (hunk 1 above)
Comment (Boundarybreaker) [reply src/main/java/com/swordglowsblue/artifice/mixin/MixinMinecraftClient.java]: Yeah, mojang got rid of the `initialize()` method in `MinecraftClient` and you can only inject into constructors at `RETURN`, so you gotta redirect or find some other way to get your code in. Fabric API had this issue too.

**Comment by arthurbambou** [comment]
Tested on 1.15 and I didn’t find any issue.

**Comment by arthurbambou** [comment]
@swordglowsblue I know you are busy but can you review this again please? I think it would be useful for some dev and mods.
